storage-proofs/src/hybrid_merkle.rs

```diff
@@ -0,0 +1,1445 @@
+    ///     2 => {third node value as beta hasher domain element}
+    ///     ...
+    ///
+    /// Layer #1 ({hasher id: beta, alpha}):
+    ///     ...
+    #[allow(clippy::needless_range_loop)]
+    pub fn pretty_print(&self) {
+        let last_beta_layer = self.beta_tree_height;
+        let first_alpha_layer = last_beta_layer;
+        let last_node_index_per_layer = self.last_node_index_per_layer();
+        let mut first_node_in_layer = 0;
+
+        for layer_index in 0..=last_beta_layer {
+            println!("\n\nLayer #{} (beta):", layer_index);
+            let last_node_in_layer = last_node_index_per_layer[layer_index];
+            for node_index in first_node_in_layer..=last_node_in_layer {
+                println!(
+                    "\t{} => {:?}",
+                    node_index,
+                    self.read_at(node_index).beta.unwrap()
+                );
+            }
+            if layer_index != last_beta_layer {
+                first_node_in_layer = last_node_in_layer + 1;
+            }
+        }
+
+        for layer_index in first_alpha_layer..=self.height {
+            println!("\n\nLayer #{} (alpha):", layer_index);
+            let last_node_in_layer = last_node_index_per_layer[layer_index];
+            for node_index in first_node_in_layer..=last_node_in_layer {
+                println!(
+                    "\t{} => {:?}",
+                    node_index,
+                    self.read_at(node_index).alpha.unwrap()
+                );
+            }
+            first_node_in_layer = last_node_in_layer + 1;
+        }
+    }
+}
+
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct HybridMerkleProof<AH, BH>
+where
+    AH: Hasher,
+    BH: Hasher,
+{
+    challenge_value: BH::Domain,
+    beta_path: Vec<(BH::Domain, bool)>,
+    alpha_path: Vec<(AH::Domain, bool)>,
+    root: AH::Domain,
+
+    #[serde(skip)]
+    _ah: PhantomData<AH>,
+
+    #[serde(skip)]
+    _bh: PhantomData<BH>,
+}
+
+impl<AH, BH> HybridMerkleProof<AH, BH>
+where
+    AH: Hasher,
+    BH: Hasher,
+{
+    pub fn new_empty(height: usize) -> Self {
+        let beta_path_len = height - ALPHA_TREE_HEIGHT;
+        let beta_path = vec![(BH::Domain::default(), false); beta_path_len];
+        let alpha_path = vec![(AH::Domain::default(), false); ALPHA_TREE_HEIGHT];
+
+        HybridMerkleProof {
+            beta_path,
+            alpha_path,
+            ..Default::default()
+        }
+    }
+
+    #[allow(clippy::explicit_counter_loop)]
+    fn calculate_beta_root(&self) -> BH::Domain {
+        let mut tree_layer = 0;
+        let (sibling, sibling_is_left) = self.beta_path[0];
+
+        let mut child = if sibling_is_left {
+            hash_node::<BH>(sibling, self.challenge_value, tree_layer)
+        } else {
+            hash_node::<BH>(self.challenge_value, sibling, tree_layer)
+        };
+
+        for (sibling, sibling_is_left) in &self.beta_path[1..] {
+            tree_layer += 1;
```

Comment:
I think that it is more explicit to use the same variable for the same value both outside of and inside of the for-loop, rather than use a constant `tree_layer = 0` above the for-loop and use the `enumerate`d value as the `tree_layer` within the for-loop